apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/RetryHandler.cs

```diff
@@ -0,0 +1,67 @@
+﻿// Copyright 2022 Google Inc. All Rights Reserved.
+// 
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// 
+//     http://www.apache.org/licenses/LICENSE-2.0
+// 
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using Google.Apis.Storage.v1;
+using Google.Apis.Http;
+using Google.Apis.Json;
+using Google.Apis.Util;
+using System;
+using System.Linq;
+using System.Net;
+using System.Net.Http;
+using System.Threading.Tasks;
+using System.Net.WebSockets;
+
+namespace Google.Cloud.Storage.V1
+{
+    /// <summary>
+    /// Handler for exponential backoff for request messages which permit retries.
+    /// This is similar to BackoffHandler in Google.Apis.Http, but that doesn't give us enough flexibility.
+    /// This handler does not enforce a maximum number of retries; that's configured via <see cref="ConfigurableMessageHandler.NumTries"/>.
+    /// </summary>
+    internal sealed class RetryHandler : IHttpUnsuccessfulResponseHandler
+    {
+        private static readonly int[] s_retriableErrorCodes = { 408, 429, 500, 502, 503, 504 };
+        private static RetryHandler s_instance = new RetryHandler();
+
+        private RetryHandler() { }
+
+        internal static void MarkAsRetriable<TResponse>(StorageBaseServiceRequest<TResponse> request) =>
+            request.AddUnsuccessfulResponseHandler(s_instance);
+
+        internal static Task<bool> IsRetriableResponse(HttpResponseMessage response)
+        {
+            return Task.FromResult(s_retriableErrorCodes.Contains(((int)response.StatusCode)));
```

Comment:
It might be worth noting in a comment that this is designed to be asynchronous in case we need to examine the response content, but for now we only need the status code.